extensions/cornerstone-dynamic-volume/src/panels/PanelGenerateImage.tsx

```diff
@@ -0,0 +1,376 @@
+        }
+      });
+      return newState;
+    });
+  };
+
+  // Establish a reference to the viewer API context
+  const { activeViewportIndex, viewports } = viewportGridService.getState();
+  const displaySetInstanceUID = viewports[0].displaySetInstanceUIDs[0];
+  const computedVolumeId = `cornerstoneStreamingImageVolume:${uuidComputedVolume.current}`;
+  const dynamicVolumeId = `${volumeLoaderScheme}:${displaySetInstanceUID}`;
+
+  useEffect(() => {
+    // ~~ Subscription
+    const evt = cornerstoneViewportService.EVENTS.VIEWPORT_DATA_CHANGED;
+    const subscriptions = [];
+
+    subscriptions.push(
+      cornerstoneViewportService.subscribe(evt, evtdetails => {
+        evtdetails.viewportData.data.forEach(volumeData => {
+          if (volumeData.volumeId.split(':')[0] === volumeLoaderScheme) {
+            setDynamicVolume(volumeData.volume);
+            uuidDynamicVolume.current = volumeData.displaySetInstanceUID;
+            const range = [1, volumeData.volume._numTimePoints];
+            setRangeValues(prevArray => [...prevArray, ...range]);
+            setTimeFramesToUse([0, 1]);
+          }
+        });
+      }).unsubscribe
+    );
+
+    return () => {
+      subscriptions.forEach(unsub => {
+        unsub();
+      });
+    };
+  }, []);
+
+  function renderGeneratedImage(displaySet) {
+    commandsManager.runCommand('setDisplaySetToGridViewports', {
+      displaySet,
+    });
+  }
+
+  // Get computed volume from cache, calculate the data across the time frames,
+  // set the scalar data to the computedVolume, and create displaySet
+  async function onGenerateImage() {
+    const computedVolumeInit = await createComputedVolume(
+      dynamicVolume.volumeId,
+      computedVolumeId
+    );
+
+    const computedVolume = cache.getVolume(computedVolumeId);
+    const dataInTime = cstUtils.dynamicVolume.generateImageFromTimeData(
+      dynamicVolume,
+      options.Operation,
+      timeFramesToUse
+    );
+
+    // Add loadStatus.loaded to computed volume and set to true
+    computedVolume.loadStatus = {};
+    computedVolume.loadStatus.loaded = true;
+
+    // Set computed scalar data to volume
+    const scalarData = computedVolume.getScalarData();
+    for (let i = 0; i < dataInTime.length; i++) {
+      scalarData[i] = dataInTime[i];
+    }
+
+    // If computed display set does not exist, create an object to be used as
+    // the displaySet. If it does exist, update the image data and vtkTexture
+    if (!computedDisplaySet) {
+      const obj = {
+        [uuidComputedVolume.current]: {
+          volumeLoaderSchema: computedVolume.volumeId.split(':')[0],
+          displaySetInstanceUID: uuidComputedVolume.current,
+          SOPClassHandlerId: SOPClassHandlerId,
+          Modality: dynamicVolume.metadata.Modality,
+          isMultiFrame: false,
+          numImageFrames: 1,
+          uid: uuidComputedVolume.current,
+        },
+      };
+      setComputedDisplaySet(obj);
```

Comment:
Why an object/map using the UUID as key if no other displaySet is added to it? That is why I thought the `displaySet` variable above should be called `displaySets`.

There is an `useEffect` below that will be called only when the setComputedDisplaySet above is called, right? Why not call `renderGeneratedImage(computedDisplaySet)` here and remove the `useEffect`?

Reply:
Added `renderGeneratedImage()` to call after creating the displaySet, removed `useEffect` that would call when `computedDisplaySet` was updated.

Reply:
As for the dictionary, I was getting issues when caching the display set when I wasn't using a dictionary